gensim/models/keyedvectors.py

```diff
@@ -558,6 +564,134 @@ def most_similar(self, positive=None, negative=None, topn=10, restrict_vocab=Non
         result = [(self.index2word[sim], float(dists[sim])) for sim in best if sim not in all_words]
         return result[:topn]
 
+    def evaluate_word_analogies_set_based(self, analogies, restrict_vocab=300000,
+    case_insensitive=True, dummy4unknown=False, topk=5):
+        """Compute performance of the model on an analogy test set using a set based method.
+
+        The accuracy is reported (printed to log and returned as a score) for each section separately,
+        plus there's one aggregate summary at the end. It implements set based 3CosAvg method.
+        It has been implemented in a similar way with the original evaluate_word_analogies method
+        which is for the 3CosAdd OR 3CosMul methods. According to 3CosAvg method we compute
+        the average of the difference between every given pair of words and then we add it
+        in the source word of the last pair trying to predict the target word of the last pair.
+        This method corresponds to the `compute-accuracy` script of the original C word2vec.
+        See also `3CosAvg <https://www.aclweb.org/anthology/C/C16/C16-1332.pdf>`_. For more
+        details see  on GitHub #2321 <https://github.com/RaRe-Technologies/gensim/pull/2321>`_.
+
+        Parameters
+        ----------
+        analogies : str
+            Path to file, where lines are tuples of words, split into sections by ": SECTION NAME" lines.
+            See `gensim/test/test_data/questions-words.txt` as example.
+        restrict_vocab : int, optional
+            Ignore all tuples containing a word not in the first `restrict_vocab` words.
+            This may be meaningful if you've sorted the model vocabulary by descending frequency (which is standard
+            in modern word embedding models).
+        case_insensitive : bool, optional
+            If True - convert all words to their uppercase form before evaluating the performance.
+            Useful to handle case-mismatch between training tokens and words in the test set.
+            In case of multiple case variants of a single word, the vector for the first occurrence
+            (also the most frequent if vocabulary is sorted) is taken.
+        dummy4unknown : bool, optional
+            If True - produce zero accuracies for tuples with out-of-vocabulary words.
+            Otherwise, these tuples are skipped entirely and not used in the evaluation.
+        topk: int, optional
+            Number of the top similar words we can consider for a successful prediction.
+
+        Returns
+        -------
+        (float, list of dict of (str, (str, str, ...., str))
+            Overall evaluation score and full lists of correct and incorrect predictions divided by sections.
```

Comment:
Can you clarify the formats? What is `dict of (str, (str, str, ...., str))`, specifically?

A concrete example here would help a lot IMO.